system/lib/wasmfs/wasmfs.cpp

```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
+      std::make_shared<OpenFileState>(0, flags, getRootDirectory());
+    return FileTable::get().add(openFile);
+  }
 
-  __wasi_size_t num = 0;
-  for (size_t i = 0; i < iovs_len; i++) {
-    const uint8_t* buf = iovs[i].buf;
-    __wasi_size_t len = iovs[i].buf_len;
+  long err;
+  std::shared_ptr<Directory> parentDir = getParent(pathParts, err);
 
-    lockedFile.read(buf, len);
-    num += len;
+  // Parent node doesn't exist.
+  if (!parentDir) {
+    return err;
   }
-  *nread = num;
-  return __WASI_ERRNO_INVAL;
+
+  auto lockedParentDir = parentDir->locked();
+
+  std::shared_ptr<File> curr =
+    lockedParentDir.getEntry(pathParts[pathParts.size() - 1]);
+
+  // The requested node was not found.
+  if (!curr) {
+    // If curr is the last element and the create flag is specified
+    if (flags & O_CREAT) {
+      // Create an empty in-memory file.
+      auto created = std::make_shared<MemoryFile>(mode);
+
+      lockedParentDir.setEntry(pathParts[pathParts.size() - 1], created);
+      auto openFile = std::make_shared<OpenFileState>(0, flags, created);
+
+      return FileTable::get().add(openFile);
+    } else {
+      return -(ENOENT);
+    }
+  }
+
+  // Fail if O_DIRECTORY is specified and pathname is not a directory
+  if (flags & O_DIRECTORY && !curr->is<Directory>()) {
+    return -(ENOTDIR);
+  }
+
+  // If the file exists and O_EXCL and O_CREAT are true
+  if (flags & O_EXCL && flags & O_CREAT) {
+    return -(EEXIST);
+  }
+
+  auto openFile = std::make_shared<OpenFileState>(0, flags, curr);
+
+  return FileTable::get().add(openFile);
 }
 
-__wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
-  int accessMode = (flags & O_ACCMODE);
-  bool canWrite = false;
+long __syscall_mkdir(long path, long mode) {
+  std::vector<std::string> pathParts = splitPath((char*)path);
 
-  if (accessMode == O_WRONLY || accessMode == O_RDWR) {
-    canWrite = true;
+  // Root directory
+  if (pathParts.empty()) {
+    return -(EEXIST);
   }
 
-  std::vector<std::string> pathParts;
+  long err;
+  std::shared_ptr<Directory> parentDir = getParent(pathParts, err);
 
-  char newPathName[strlen((char*)pathname) + 1];
-  strcpy(newPathName, (char*)pathname);
+  if (!parentDir) {
+    // parent node doesn't exist
+    return err;
+  }
+
+  auto lockedParentDir = parentDir->locked();
+
+  std::shared_ptr<File> curr =
+    lockedParentDir.getEntry(pathParts[pathParts.size() - 1]);
 
-  // TODO: Support relative paths. i.e. specify cwd if path is relative.
-  // TODO: Other path parsing edge cases.
-  char* current;
+  // Check if the request directory already exists.
+  if (curr) {
+    return -(EEXIST);
+  } else {
+    // Create an empty in-memory file.
+    auto created = std::make_shared<Directory>(mode);
 
-  current = strtok(newPathName, "/\n");
-  while (current != NULL) {
-    pathParts.push_back(current);
-    current = strtok(NULL, "/\n");
+    lockedParentDir.setEntry(pathParts[pathParts.size() - 1], created);
```

Comment:
`pathParts[pathParts.size() - 1]` appears twice and is not very clear by itself. How about using a temp var for it,
```
auto base = pathParts[pathParts.size() - 1];
```
(or "name", "basename", etc. - "base name" is what python calls it at least, not sure if there's not a better term).